### PR TITLE
kubeadm: Add etcd client unit tests

### PR DESCRIPTION
**Pull request by dlipovetsky**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
As the result of the discussion that started with https://github.com/kubernetes/kubernetes/pull/117724#issuecomment-1533322854, I created a unit test to exercise kubeadm's etcd client, which wraps the upstream etcd client. In order to create the test, I made kubdeadm consume the upstream etcd client via an interface.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/util/etcd/etcd_test.go [hunk 1]
```diff
@@ -351,3 +352,141 @@ func TestGetRawEtcdEndpointsFromPodAnnotationWithoutRetry(t *testing.T) {
...
+// MemberList lists the current cluster membership.
+func (f *fakeEtcdClient) MemberList(_ context.Context) (*clientv3.MemberListResponse, error) {
+	return &clientv3.MemberListResponse{
+		Members: f.members,
+	}, nil
+}
+
+// MemberAdd adds a new member into the cluster.
+func (f *fakeEtcdClient) MemberAdd(_ context.Context, peerAddrs []string) (*clientv3.MemberAddResponse, error) {
+	return nil, fmt.Errorf("not implemented")
```
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd_test.go:381]: these should be `github.com/pkg/errors.New(..)`
for consistency.

EDIT: we can define a file-local not implemented error var and return it where needed.

**Review by neolit123 — COMMENTED** [review]
/approve
thanks @dlipovetsky these additions seems fine.

File: cmd/kubeadm/app/util/etcd/etcd_test.go [hunk 2]
```diff
@@ -351,3 +352,141 @@ func TestGetRawEtcdEndpointsFromPodAnnotationWithoutRetry(t *testing.T) {
 		})
 	}
 }
+
+type fakeEtcdClient struct {
```
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd_test.go:356]: we can move this near the top of the file.

File: cmd/kubeadm/app/util/etcd/etcd.go [hunk 3]
```diff
@@ -82,52 +88,46 @@ type etcdClient interface {
...
-		tlsConfig, err := tlsInfo.ClientConfig()
+		tlsConfig, err = tlsInfo.ClientConfig()
 		if err != nil {
 			return nil, err
 		}
-		client.TLS = tlsConfig
 	}
 
-	client.newEtcdClient = func() (etcdClient, error) {
+	client.newEtcdClient = func(endpoints []string) (etcdClient, error) {
```
Comment (chendave) [inline cmd/kubeadm/app/util/etcd/etcd.go:123]: what's the benefit of passing the `endpoints` here vs. the original way to set the field of  `Endpoints` by `client.Endpoints`?

File: cmd/kubeadm/app/util/etcd/etcd_test.go [hunk 4]
```diff
@@ -17,27 +17,28 @@ limitations under the License.
...
 
 import (
+	"context"
 	"fmt"
 	"reflect"
 	"strconv"
 	"testing"
 
 	"github.com/pkg/errors"
-
```
Comment (chendave) [inline cmd/kubeadm/app/util/etcd/etcd_test.go:26]: pls revert the change here, the line here is intentionally for the separation of different kinds of packages.

File: cmd/kubeadm/app/util/etcd/etcd.go [hunk 5]
```diff
@@ -66,10 +66,41 @@ type ClusterInterrogator interface {
...
+	Endpoints() []string
+
+	// Sync synchronizes client's endpoints with the known endpoints from the etcd membership.
+	Sync(ctx context.Context) error
+
+	// Status gets the status of the endpoint.
+	Status(ctx context.Context, endpoint string) (*clientv3.StatusResponse, error)
+
+	// Close shuts down the client's etcd connections.
+	Close() error
```
Comment (chendave) [inline cmd/kubeadm/app/util/etcd/etcd.go]: Can I ask how do you select the methods which should be implemented by the interface? i.e. why the `Close()` is here and `MemberUpdate` is not?

File: cmd/kubeadm/app/util/etcd/etcd.go [hunk 6]
```diff
@@ -67,6 +67,12 @@ type ClusterInterrogator interface {
 }
 
 type etcdClient interface {
+	// Close shuts down the client's etcd connections.
+	Close() error
+
+	// Endpoints lists the registered endpoints for the client.
+	Endpoints() []string
```
[the remark this answers is not in the document]
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd.go:74]: My fault! These changes ended up in the second commit by mistake. I'll move them into the first, as you ask.

File: cmd/kubeadm/app/util/etcd/etcd.go (hunk 3 above)
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd.go:123]: Good question.

I [instantiate](https://github.com/kubernetes/kubernetes/blob/44819c18482ff1fae243df4210488b824e838d38/cmd/kubeadm/app/util/etcd/etcd.go#L123) `client.newEtcdClient` using an anonymous closure.

Originally, I _did_ reference the struct member `Endpoints` from the closure. However, the `Endpoints` member [is updated by the `Sync` receiver](https://github.com/kubernetes/kubernetes/blob/44819c18482ff1fae243df4210488b824e838d38/cmd/kubeadm/app/util/etcd/etcd.go#L262). I felt that it was not clear to the reader that the closure would read the up-to-date value of `Endpoints`.

I thought that explicitly passing the value to the closure would make it  clear that the closure uses the up-to-date value of `Endpoints`.

File: cmd/kubeadm/app/util/etcd/etcd.go (hunk 5 above)
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd.go]: Good question.

I selected the methods that _have_ to be in the interface, because they are called at least once by kubeadm. The `MemberUpdate` method is not called.

File: cmd/kubeadm/app/util/etcd/etcd.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd.go:74]: Fixed by https://github.com/kubernetes/kubernetes/pull/117792/commits/55399ca366a4a8a58fe485240444849474dea098 and https://github.com/kubernetes/kubernetes/pull/117792/commits/c6fd36daacf262589a39ea3be6db3f6a4532c00b

File: cmd/kubeadm/app/util/etcd/etcd_test.go [hunk 7]
```diff
@@ -17,27 +17,28 @@ limitations under the License.
...
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
 	clientsetfake "k8s.io/client-go/kubernetes/fake"
 	clienttesting "k8s.io/client-go/testing"
-
```
[the remark this answers is not in the document]
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd_test.go]: Fixed by https://github.com/kubernetes/kubernetes/pull/117792/commits/f826a9eef81b0f463a5726f70497cce8fdba617a

File: cmd/kubeadm/app/util/etcd/etcd_test.go (hunk 4 above)
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd_test.go:26]: Fixed by https://github.com/kubernetes/kubernetes/pull/117792/commits/f826a9eef81b0f463a5726f70497cce8fdba617a

File: cmd/kubeadm/app/util/etcd/etcd_test.go (hunk 2 above)
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd_test.go]: Fixed by https://github.com/kubernetes/kubernetes/pull/117792/commits/ee828f7c90caa9b25edd7499997d8a3d50478105

File: cmd/kubeadm/app/util/etcd/etcd_test.go (hunk 1 above)
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd_test.go]: Fixed by https://github.com/kubernetes/kubernetes/pull/117792/commits/9010304aec81062e34363fa69d653bdb03a86337

**Comment by dlipovetsky** [comment]
Adding a hold so I can squash the fix-up commits before merging :smile:  

/hold

File: cmd/kubeadm/app/util/etcd/etcd.go (hunk 5 above)
Comment (neolit123) [reply cmd/kubeadm/app/util/etcd/etcd.go]: perhaps we should just add all methods in the interface

**Comment by neolit123** [comment]
please remember to squash the fuxup commits before lgtm

**Comment by dlipovetsky** [comment]
> please remember to squash the fuxup commits before lgtm

Oh? I used fixup commits to make it easier for reviewers to see the changes, and my plan was to squash the commits _after_ I got an /lgtm, and then to cancel the /hold after squashing.

**Comment by neolit123** [comment]
> > please remember to squash the fuxup commits before lgtm
> 
> Oh? I used fixup commits to make it easier for reviewers to see the changes, and my plan was to squash the commits _after_ I got an /lgtm, and then to cancel the /hold after squashing.

i think i saw a discussion about the k8s CI feature to preserve lgtm after squash, but not sure if it has landed yet. if yes, great!

**Comment by dlipovetsky** [comment]
> > > please remember to squash the fuxup commits before lgtm
> > 
> > 
> > Oh? I used fixup commits to make it easier for reviewers to see the changes, and my plan was to squash the commits _after_ I got an /lgtm, and then to cancel the /hold after squashing.
> 
> i think i saw a discussion about the k8s CI feature to preserve lgtm after squash, but not sure if it has landed yet. if yes, great!

I think it landed in https://github.com/kubernetes/test-infra/pull/22663

File: cmd/kubeadm/app/util/etcd/etcd.go (hunk 5 above)
Comment (dlipovetsky) [reply cmd/kubeadm/app/util/etcd/etcd.go]: > perhaps we should just add all methods in the interface

I wouldn't want to include any methods not used by the "real" code.

The Go convention, as far as I understand it, is every consumer (e.g. every package) defines their own interface, and includes only the methods they call. There are exceptions to this convention, where the producer exports an Interface for consumers to import, but that's not the case here.

Comment (neolit123) [reply cmd/kubeadm/app/util/etcd/etcd.go]: ok, that's fair

**Comment by dlipovetsky** [comment]
Thanks for the review, @chendave!

File: cmd/kubeadm/app/util/etcd/etcd.go (hunk 3 above)
Comment (chendave) [reply cmd/kubeadm/app/util/etcd/etcd.go:123]: Ack.

**Review by chendave — COMMENTED** [review]
/lgtm
thanks!

pls quash by the way.

**Comment by dlipovetsky** [comment]
Somehow my local rebase pulled in some extra commits, although `git diff origin/kubeadm-etcd-client-refactor` showed no difference just before I force-pushed :facepalm:. Since all the labels were removed anyway, I just rebased on latest master.

**Comment by chendave** [comment]
still lgtm
/lgtm

**Comment by dlipovetsky** [comment]
Since I squashed the fixups, I'll remove the hold.

Once this merges, I'll need to rebase #117724 and update the unit test, but that's actually a positive! :smile: 

/hold cancel
